src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -1289,21 +1289,32 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
                     );
                 };
             }
-            StatementKind::AscribeUserType(ref place, variance, box ref c_ty) => {
+            StatementKind::AscribeUserType(ascription) => {
+                let UserTypeAscriptionData {
+                    ref place,
+                    variance,
+                    box ref user_type_projection,
+                    span,
+                } = self.mir.user_type_ascriptions[ascription];
+
+                debug!(
+                    "check_stmt: place={:?} variance={:?} user_type_projection={:?} span={:?}",
+                    place, variance, user_type_projection, span,
+                );
                 let place_ty = place.ty(mir, tcx).to_ty(tcx);
                 if let Err(terr) = self.relate_type_and_user_type(
                     place_ty,
                     variance,
-                    c_ty,
-                    Locations::All(stmt.source_info.span),
+                    user_type_projection,
+                    Locations::All(span),
                     ConstraintCategory::TypeAnnotation,
                 ) {
                     span_mirbug!(
                         self,
                         stmt,
                         "bad type assert ({:?} <: {:?}): {:?}",
                         place_ty,
-                        c_ty,
+                        user_type_projection,
```

Comment:
I suspect we will want to extend the `TypeVerifier` stuff:

https://github.com/rust-lang/rust/blob/6f244c9011eb01ae4d167b631ba5e3e163bafa72/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L268

well, actually, just adding to the visitor should be ok. The main thing is that we want

https://github.com/rust-lang/rust/blob/6f244c9011eb01ae4d167b631ba5e3e163bafa72/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L275

to be invoked.

Reply:
Resolving this as all the code has changed.